cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java

```diff
@@ -524,46 +609,26 @@ private MetadataEntry convertRow(Row row) {
-   * @param searchQuery the search query, which could be of two forms: [key]:[value] or just [value] and can have '*'
-   *                    at the end for a prefix search
-   * @param types the {@link EntityTypeSimpleName} to restrict the search to, if empty all types are searched
-   * @param sortInfo the {@link SortInfo} to sort the results by
-   * @param offset index to start with in the search results. To return results from the beginning, pass {@code 0}.
-   *               Only applies when #sortInfo is not {@link SortInfo#DEFAULT}
-   * @param limit number of results to return, starting from #offset. To return all, pass {@link Integer#MAX_VALUE}.
-   *              Only applies when #sortInfo is not {@link SortInfo#DEFAULT}
-   * @param numCursors number of cursors to return in the response. A cursor identifies the first index of the
-   *                   next page for pagination purposes. Only applies when #sortInfo is not {@link SortInfo#DEFAULT}.
-   *                   Defaults to {@code 0}
-   * @param cursor the cursor that acts as the starting index for the requested page. This is only applicable when
-   *               #sortInfo is not {@link SortInfo#DEFAULT}. If offset is also specified, it is applied starting at
-   *               the cursor. If {@code null}, the first row is used as the cursor
-   * @param showHidden boolean which specifies whether to display hidden entities (entity whose name start with "_")
-   *                    or not.
-   * @param entityScope a set which specifies which scope of entities to display.
+   * @param request the search request
    * @return a {@link SearchResults} object containing a list of {@link MetadataEntry} containing each matching
    *         {@link MetadataEntity} with its associated metadata. It also optionally contains a list of cursors
    *         for subsequent queries to start with, if the specified #sortInfo is not {@link SortInfo#DEFAULT}.
    */
-  public SearchResults search(String namespaceId, String searchQuery, Set<EntityTypeSimpleName> types,
-                              SortInfo sortInfo, int offset, int limit, int numCursors, @Nullable String cursor,
-                              boolean showHidden, Set<EntityScope> entityScope) throws BadRequestException {
-    if (!SortInfo.DEFAULT.equals(sortInfo)) {
-      if (!"*".equals(searchQuery)) {
-        throw new BadRequestException("Cannot search with non-default sort with any query other than '*'");
-      }
-      return searchByCustomIndex(namespaceId, types, sortInfo, offset, limit, numCursors, cursor, showHidden,
-                                 entityScope);
+  public SearchResults search(SearchRequest request) throws BadRequestException {
+    if (SortInfo.DEFAULT.equals(request.getSortInfo())) {
+      return searchByDefaultIndex(request);
     }
-    return searchByDefaultIndex(namespaceId, searchQuery, types, showHidden, entityScope);
+
+    if (!"*".equals(request.getQuery())) {
```

Comment:
not a part of this PR but can be move these wildcard checks to SearchRequest itself to encapsulate the *

Reply:
I was thinking about that but decided to keep it here because this is actually a limitation of the dataset, but not an invalid search request in general. If we moved to some other search system, it might be possible.